Could not translate Error: Request failed with status code 429 + Cannot find module '@typescript-eslint/parser'

Hi,

the issue is the same for Windows 10 nodejs v13.9.0 and Linux nodejs v10.18.0 (Raspberry Pi). The error is for "Windows Admin", "Windows Normal User", "Linux Sudo User" and "Linux Normal User".

The following is the issue:

```sh
pi@pi:/opt/iobroker/node_modules/iobroker.natsclient $ sudo npx @iobroker/create-adapter
npx: installed 173 in 60.269s

=====================================================
   Welcome to the ioBroker adapter creator v1.24.0!
=====================================================

You can cancel at any point by pressing Ctrl+C.

Let's get started with a few questions about your project!
✔ Please enter the name of your project: · testpub
✔ Which title should be shown in the admin UI? · test publisher
✔ Please enter a short description: · test publisher (client)
✔ Enter some keywords (separated by commas) to describe your project: · test,streaming,messaging,broker
✔ If you have any contributors, please enter their names (seperated by commas): · email@email.com

Nice! Let's get technical...
✔ How detailed do you want to configure your project? · no
✔ Which features should your project contain? · adapter
✔ Which category does your adapter fall into? · protocols
✔ From where will the adapter get its data? · local
✔ How will the adapter receive its data? · poll
✔ Which language do you want to use to code the adapter? · JavaScript
✔ Which of the following tools do you want to use? · ESLint, type checking
✔ Do you prefer tab or space indentation? · Tab
✔ Do you prefer double or single quotes? · double

Almost done! Just a few administrative details...
✔ Please enter your name (or nickname): · mbecker
✔ What's your name/org on GitHub? · mbecker
✔ What's your email address? · email@email.com
✔ Which license should be used for your project? · MIT License

That's it. Please wait a minute while I get this working...
[1/2] Generating files...
Could not translate to "ru": Error: Request failed with status code 429

Could not translate to "zh-cn": Error: Request failed with status code 429

Could not translate to "it": Error: Request failed with status code 429

Could not translate to "pt": Error: Request failed with status code 429

Could not translate to "pl": Error: Request failed with status code 429

Could not translate to "de": Error: Request failed with status code 429

Could not translate to "es": Error: Request failed with status code 429

Could not translate to "de": Error: Request failed with status code 429

Could not translate to "fr": Error: Request failed with status code 429

Could not translate to "nl": Error: Request failed with status code 429

Could not translate to "nl": Error: Request failed with status code 429

Could not translate to "ru": Error: Request failed with status code 429

Could not translate to "de": Error: Request failed with status code 429

Could not translate to "ru": Error: Request failed with status code 429

Could not translate to "it": Error: Request failed with status code 429

Could not translate to "pl": Error: Request failed with status code 429

Could not translate to "pt": Error: Request failed with status code 429

Could not translate to "fr": Error: Request failed with status code 429

Could not translate to "zh-cn": Error: Request failed with status code 429

Could not translate to "es": Error: Request failed with status code 429

Could not translate to "pt": Error: Request failed with status code 429

Could not translate to "nl": Error: Request failed with status code 429

Could not translate to "fr": Error: Request failed with status code 429

Could not translate to "it": Error: Request failed with status code 429

Could not translate to "es": Error: Request failed with status code 429

Could not translate to "pl": Error: Request failed with status code 429

Could not translate to "zh-cn": Error: Request failed with status code 429

Could not translate to "de": Error: Request failed with status code 429

Could not translate to "ru": Error: Request failed with status code 429

Could not translate to "pt": Error: Request failed with status code 429

Could not translate to "nl": Error: Request failed with status code 429

Could not translate to "fr": Error: Request failed with status code 429

Could not translate to "it": Error: Request failed with status code 429

Could not translate to "es": Error: Request failed with status code 429

Could not translate to "pl": Error: Request failed with status code 429

Could not translate to "zh-cn": Error: Request failed with status code 429

(node:6578) UnhandledPromiseRejectionWarning: Error: Cannot find module '@typescript-eslint/parser'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.tsFixQuotes (/root/.npm/_npx/6578/lib/node_modules/@iobroker/create-adapter/build/src/lib/tools.js:260:54)
    at files.map (/root/.npm/_npx/6578/lib/node_modules/@iobroker/create-adapter/build/src/lib/createAdapter.js:87:41)
    at Array.map (<anonymous>)
    at formatFiles (/root/.npm/_npx/6578/lib/node_modules/@iobroker/create-adapter/build/src/lib/createAdapter.js:70:18)
    at Object.createFiles (/root/.npm/_npx/6578/lib/node_modules/@iobroker/create-adapter/build/src/lib/createAdapter.js:52:12)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:6578) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 38)
(node:6578) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

P.s. I'm trying to create the adapter in the directory "node_modules" from iobroker to start the development process with Visual Studio SSH Client. With Windows I'm trying to use the adapter on my local PC (same error as stated above).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ## __WORK IN PROGRESS__
 * (AlCalzone) Update the CI scripts (including Node.js 14.x)
+* (AlCalzone) Print better error message when translations hit the rate limiter
+* (AlCalzone) Synchronize translation implementation in TS with the JS one
 
 ## 1.24.2 (2020-04-27)
 * (AlCalzone) Fixed compatibility with `@types/iobroker@3.0.6`

--- a/src/lib/tools.ts
+++ b/src/lib/tools.ts
@@ -237,7 +237,13 @@ export async function translateText(
 			}
 			error(`Invalid response for translate request`);
 		} catch (e) {
-			error(`Could not translate to "${targetLang}": ${e}`);
+			if (e.response?.status === 429) {
+				error(
+					`Could not translate to "${targetLang}": Rate-limited by Google Translate`,
+				);
+			} else {
+				error(`Could not translate to "${targetLang}": ${e}`);
+			}
 		}
 		return text;
 	}

--- a/templates/lib/tools.raw.js
+++ b/templates/lib/tools.raw.js
@@ -34,12 +34,12 @@ async function translateText(text, targetLang, yandexApiKey) {
 	if (targetLang === "en") {
 		return text;
 	} else if (!text) {
-		return '';
+		return "";
 	}
 	if (yandexApiKey) {
-		return await translateYandex(text, targetLang, yandexApiKey);
+		return translateYandex(text, targetLang, yandexApiKey);
 	} else {
-		return await translateGoogle(text, targetLang);
+		return translateGoogle(text, targetLang);
 	}
 }
 
@@ -47,7 +47,7 @@ async function translateText(text, targetLang, yandexApiKey) {
  * Translates text with Yandex API
  * @param {string} text The text to translate
  * @param {string} targetLang The target languate
- * @param {string} [apiKey] The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ * @param {string} apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  * @returns {Promise<string>}
  */
 async function translateYandex(text, targetLang, apiKey) {
@@ -57,8 +57,8 @@ async function translateYandex(text, targetLang, apiKey) {
 	try {
 		const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
 		const response = await axios({url, timeout: 15000});
-		if (response.data && response.data["text"]) {
-			return response.data["text"][0];
+		if (response.data && response.data.text && isArray(response.data.text)) {
+			return response.data.text[0];
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
@@ -82,7 +82,13 @@ async function translateGoogle(text, targetLang) {
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
-		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		if (e.response && e.response.status === 429) {
+			throw new Error(
+				`Could not translate to "${targetLang}": Rate-limited by Google Translate`
+			);
+		} else {
+			throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		}
 	}
 }
 

--- a/templates/src/lib/tools.raw.ts
+++ b/templates/src/lib/tools.raw.ts
@@ -25,18 +25,64 @@ export function isArray(it: any): it is any[] {
  * Translates text using the Google Translate API
  * @param text The text to translate
  * @param targetLang The target languate
+ * @param yandexApiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  */
-export async function translateText(text: string, targetLang: string): Promise<string> {
-	if (targetLang === "en") return text;
+export async function translateText(text: string, targetLang: string, yandexApiKey?: string): Promise<string> {
+	if (targetLang === "en") {
+		return text;
+	} else if (!text) {
+		return "";
+	}
+	if (yandexApiKey) {
+		return translateYandex(text, targetLang, yandexApiKey);
+	} else {
+		return translateGoogle(text, targetLang);
+	}
+}
+
+/**
+ * Translates text with Yandex API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ * @param apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ */
+async function translateYandex(text: string, targetLang: string, apiKey: string): Promise<string> {
+	if (targetLang === "zh-cn") {
+		targetLang = "zh";
+	}
+	try {
+		const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
+		const response = await axios({url, timeout: 15000});
+		if (isArray(response.data?.text)) {
+			return response.data.text[0];
+		}
+		throw new Error("Invalid response for translate request");
+	} catch (e) {
+		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+	}
+}
+
+/**
+ * Translates text with Google API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ */
+async function translateGoogle(text: string, targetLang: string): Promise<string> {
 	try {
 		const url = `http://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=${targetLang}&dt=t&q=${encodeURIComponent(text)}&ie=UTF-8&oe=UTF-8`;
-		const response = await axios({url, timeout: 5000});
+		const response = await axios({url, timeout: 15000});
 		if (isArray(response.data)) {
 			// we got a valid response
 			return response.data[0][0][0];
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
-		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		if (e.response?.status === 429) {
+			throw new Error(
+				`Could not translate to "${targetLang}": Rate-limited by Google Translate`
+			);
+		} else {
+			throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		}
 	}
 }

--- a/test/baselines/TS_SingleQuotes/src/lib/tools.ts
+++ b/test/baselines/TS_SingleQuotes/src/lib/tools.ts
@@ -25,18 +25,64 @@ export function isArray(it: any): it is any[] {
  * Translates text using the Google Translate API
  * @param text The text to translate
  * @param targetLang The target languate
+ * @param yandexApiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  */
-export async function translateText(text: string, targetLang: string): Promise<string> {
-	if (targetLang === 'en') return text;
+export async function translateText(text: string, targetLang: string, yandexApiKey?: string): Promise<string> {
+	if (targetLang === 'en') {
+		return text;
+	} else if (!text) {
+		return '';
+	}
+	if (yandexApiKey) {
+		return translateYandex(text, targetLang, yandexApiKey);
+	} else {
+		return translateGoogle(text, targetLang);
+	}
+}
+
+/**
+ * Translates text with Yandex API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ * @param apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ */
+async function translateYandex(text: string, targetLang: string, apiKey: string): Promise<string> {
+	if (targetLang === 'zh-cn') {
+		targetLang = 'zh';
+	}
+	try {
+		const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
+		const response = await axios({url, timeout: 15000});
+		if (isArray(response.data?.text)) {
+			return response.data.text[0];
+		}
+		throw new Error('Invalid response for translate request');
+	} catch (e) {
+		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+	}
+}
+
+/**
+ * Translates text with Google API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ */
+async function translateGoogle(text: string, targetLang: string): Promise<string> {
 	try {
 		const url = `http://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=${targetLang}&dt=t&q=${encodeURIComponent(text)}&ie=UTF-8&oe=UTF-8`;
-		const response = await axios({url, timeout: 5000});
+		const response = await axios({url, timeout: 15000});
 		if (isArray(response.data)) {
 			// we got a valid response
 			return response.data[0][0][0];
 		}
 		throw new Error('Invalid response for translate request');
 	} catch (e) {
-		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		if (e.response?.status === 429) {
+			throw new Error(
+				`Could not translate to "${targetLang}": Rate-limited by Google Translate`
+			);
+		} else {
+			throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		}
 	}
 }

--- a/test/baselines/adapter_JS_ES6Class_ESLint_TypeChecking_Spaces_SingleQuotes_LGPLv3/lib/tools.js
+++ b/test/baselines/adapter_JS_ES6Class_ESLint_TypeChecking_Spaces_SingleQuotes_LGPLv3/lib/tools.js
@@ -37,9 +37,9 @@ async function translateText(text, targetLang, yandexApiKey) {
         return '';
     }
     if (yandexApiKey) {
-        return await translateYandex(text, targetLang, yandexApiKey);
+        return translateYandex(text, targetLang, yandexApiKey);
     } else {
-        return await translateGoogle(text, targetLang);
+        return translateGoogle(text, targetLang);
     }
 }
 
@@ -47,7 +47,7 @@ async function translateText(text, targetLang, yandexApiKey) {
  * Translates text with Yandex API
  * @param {string} text The text to translate
  * @param {string} targetLang The target languate
- * @param {string} [apiKey] The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ * @param {string} apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  * @returns {Promise<string>}
  */
 async function translateYandex(text, targetLang, apiKey) {
@@ -57,8 +57,8 @@ async function translateYandex(text, targetLang, apiKey) {
     try {
         const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
         const response = await axios({url, timeout: 15000});
-        if (response.data && response.data['text']) {
-            return response.data['text'][0];
+        if (response.data && response.data.text && isArray(response.data.text)) {
+            return response.data.text[0];
         }
         throw new Error('Invalid response for translate request');
     } catch (e) {
@@ -82,7 +82,13 @@ async function translateGoogle(text, targetLang) {
         }
         throw new Error('Invalid response for translate request');
     } catch (e) {
-        throw new Error(`Could not translate to "${targetLang}": ${e}`);
+        if (e.response && e.response.status === 429) {
+            throw new Error(
+                `Could not translate to "${targetLang}": Rate-limited by Google Translate`
+            );
+        } else {
+            throw new Error(`Could not translate to "${targetLang}": ${e}`);
+        }
     }
 }
 

--- a/test/baselines/adapter_JS_ESLint_TypeChecking_Spaces_SingleQuotes_LGPLv3/lib/tools.js
+++ b/test/baselines/adapter_JS_ESLint_TypeChecking_Spaces_SingleQuotes_LGPLv3/lib/tools.js
@@ -37,9 +37,9 @@ async function translateText(text, targetLang, yandexApiKey) {
         return '';
     }
     if (yandexApiKey) {
-        return await translateYandex(text, targetLang, yandexApiKey);
+        return translateYandex(text, targetLang, yandexApiKey);
     } else {
-        return await translateGoogle(text, targetLang);
+        return translateGoogle(text, targetLang);
     }
 }
 
@@ -47,7 +47,7 @@ async function translateText(text, targetLang, yandexApiKey) {
  * Translates text with Yandex API
  * @param {string} text The text to translate
  * @param {string} targetLang The target languate
- * @param {string} [apiKey] The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ * @param {string} apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  * @returns {Promise<string>}
  */
 async function translateYandex(text, targetLang, apiKey) {
@@ -57,8 +57,8 @@ async function translateYandex(text, targetLang, apiKey) {
     try {
         const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
         const response = await axios({url, timeout: 15000});
-        if (response.data && response.data['text']) {
-            return response.data['text'][0];
+        if (response.data && response.data.text && isArray(response.data.text)) {
+            return response.data.text[0];
         }
         throw new Error('Invalid response for translate request');
     } catch (e) {
@@ -82,7 +82,13 @@ async function translateGoogle(text, targetLang) {
         }
         throw new Error('Invalid response for translate request');
     } catch (e) {
-        throw new Error(`Could not translate to "${targetLang}": ${e}`);
+        if (e.response && e.response.status === 429) {
+            throw new Error(
+                `Could not translate to "${targetLang}": Rate-limited by Google Translate`
+            );
+        } else {
+            throw new Error(`Could not translate to "${targetLang}": ${e}`);
+        }
     }
 }
 

--- a/test/baselines/adapter_TS_ES6Class_ESLint_Tabs_DoubleQuotes_MIT/src/lib/tools.ts
+++ b/test/baselines/adapter_TS_ES6Class_ESLint_Tabs_DoubleQuotes_MIT/src/lib/tools.ts
@@ -25,18 +25,64 @@ export function isArray(it: any): it is any[] {
  * Translates text using the Google Translate API
  * @param text The text to translate
  * @param targetLang The target languate
+ * @param yandexApiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  */
-export async function translateText(text: string, targetLang: string): Promise<string> {
-	if (targetLang === "en") return text;
+export async function translateText(text: string, targetLang: string, yandexApiKey?: string): Promise<string> {
+	if (targetLang === "en") {
+		return text;
+	} else if (!text) {
+		return "";
+	}
+	if (yandexApiKey) {
+		return translateYandex(text, targetLang, yandexApiKey);
+	} else {
+		return translateGoogle(text, targetLang);
+	}
+}
+
+/**
+ * Translates text with Yandex API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ * @param apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ */
+async function translateYandex(text: string, targetLang: string, apiKey: string): Promise<string> {
+	if (targetLang === "zh-cn") {
+		targetLang = "zh";
+	}
+	try {
+		const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
+		const response = await axios({url, timeout: 15000});
+		if (isArray(response.data?.text)) {
+			return response.data.text[0];
+		}
+		throw new Error("Invalid response for translate request");
+	} catch (e) {
+		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+	}
+}
+
+/**
+ * Translates text with Google API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ */
+async function translateGoogle(text: string, targetLang: string): Promise<string> {
 	try {
 		const url = `http://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=${targetLang}&dt=t&q=${encodeURIComponent(text)}&ie=UTF-8&oe=UTF-8`;
-		const response = await axios({url, timeout: 5000});
+		const response = await axios({url, timeout: 15000});
 		if (isArray(response.data)) {
 			// we got a valid response
 			return response.data[0][0][0];
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
-		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		if (e.response?.status === 429) {
+			throw new Error(
+				`Could not translate to "${targetLang}": Rate-limited by Google Translate`
+			);
+		} else {
+			throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		}
 	}
 }

--- a/test/baselines/adapter_TS_ESLint_Tabs_DoubleQuotes_MIT/src/lib/tools.ts
+++ b/test/baselines/adapter_TS_ESLint_Tabs_DoubleQuotes_MIT/src/lib/tools.ts
@@ -25,18 +25,64 @@ export function isArray(it: any): it is any[] {
  * Translates text using the Google Translate API
  * @param text The text to translate
  * @param targetLang The target languate
+ * @param yandexApiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  */
-export async function translateText(text: string, targetLang: string): Promise<string> {
-	if (targetLang === "en") return text;
+export async function translateText(text: string, targetLang: string, yandexApiKey?: string): Promise<string> {
+	if (targetLang === "en") {
+		return text;
+	} else if (!text) {
+		return "";
+	}
+	if (yandexApiKey) {
+		return translateYandex(text, targetLang, yandexApiKey);
+	} else {
+		return translateGoogle(text, targetLang);
+	}
+}
+
+/**
+ * Translates text with Yandex API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ * @param apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ */
+async function translateYandex(text: string, targetLang: string, apiKey: string): Promise<string> {
+	if (targetLang === "zh-cn") {
+		targetLang = "zh";
+	}
+	try {
+		const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
+		const response = await axios({url, timeout: 15000});
+		if (isArray(response.data?.text)) {
+			return response.data.text[0];
+		}
+		throw new Error("Invalid response for translate request");
+	} catch (e) {
+		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+	}
+}
+
+/**
+ * Translates text with Google API
+ * @param text The text to translate
+ * @param targetLang The target languate
+ */
+async function translateGoogle(text: string, targetLang: string): Promise<string> {
 	try {
 		const url = `http://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=${targetLang}&dt=t&q=${encodeURIComponent(text)}&ie=UTF-8&oe=UTF-8`;
-		const response = await axios({url, timeout: 5000});
+		const response = await axios({url, timeout: 15000});
 		if (isArray(response.data)) {
 			// we got a valid response
 			return response.data[0][0][0];
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
-		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		if (e.response?.status === 429) {
+			throw new Error(
+				`Could not translate to "${targetLang}": Rate-limited by Google Translate`
+			);
+		} else {
+			throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		}
 	}
 }

--- a/test/baselines/vis_Widget/lib/tools.js
+++ b/test/baselines/vis_Widget/lib/tools.js
@@ -37,9 +37,9 @@ async function translateText(text, targetLang, yandexApiKey) {
 		return "";
 	}
 	if (yandexApiKey) {
-		return await translateYandex(text, targetLang, yandexApiKey);
+		return translateYandex(text, targetLang, yandexApiKey);
 	} else {
-		return await translateGoogle(text, targetLang);
+		return translateGoogle(text, targetLang);
 	}
 }
 
@@ -47,7 +47,7 @@ async function translateText(text, targetLang, yandexApiKey) {
  * Translates text with Yandex API
  * @param {string} text The text to translate
  * @param {string} targetLang The target languate
- * @param {string} [apiKey] The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ * @param {string} apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  * @returns {Promise<string>}
  */
 async function translateYandex(text, targetLang, apiKey) {
@@ -57,8 +57,8 @@ async function translateYandex(text, targetLang, apiKey) {
 	try {
 		const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
 		const response = await axios({url, timeout: 15000});
-		if (response.data && response.data["text"]) {
-			return response.data["text"][0];
+		if (response.data && response.data.text && isArray(response.data.text)) {
+			return response.data.text[0];
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
@@ -82,7 +82,13 @@ async function translateGoogle(text, targetLang) {
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
-		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		if (e.response && e.response.status === 429) {
+			throw new Error(
+				`Could not translate to "${targetLang}": Rate-limited by Google Translate`
+			);
+		} else {
+			throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		}
 	}
 }
 

--- a/test/baselines/vis_Widget_Travis/lib/tools.js
+++ b/test/baselines/vis_Widget_Travis/lib/tools.js
@@ -37,9 +37,9 @@ async function translateText(text, targetLang, yandexApiKey) {
 		return "";
 	}
 	if (yandexApiKey) {
-		return await translateYandex(text, targetLang, yandexApiKey);
+		return translateYandex(text, targetLang, yandexApiKey);
 	} else {
-		return await translateGoogle(text, targetLang);
+		return translateGoogle(text, targetLang);
 	}
 }
 
@@ -47,7 +47,7 @@ async function translateText(text, targetLang, yandexApiKey) {
  * Translates text with Yandex API
  * @param {string} text The text to translate
  * @param {string} targetLang The target languate
- * @param {string} [apiKey] The yandex API key. You can create one for free at https://translate.yandex.com/developers
+ * @param {string} apiKey The yandex API key. You can create one for free at https://translate.yandex.com/developers
  * @returns {Promise<string>}
  */
 async function translateYandex(text, targetLang, apiKey) {
@@ -57,8 +57,8 @@ async function translateYandex(text, targetLang, apiKey) {
 	try {
 		const url = `https://translate.yandex.net/api/v1.5/tr.json/translate?key=${apiKey}&text=${encodeURIComponent(text)}&lang=en-${targetLang}`;
 		const response = await axios({url, timeout: 15000});
-		if (response.data && response.data["text"]) {
-			return response.data["text"][0];
+		if (response.data && response.data.text && isArray(response.data.text)) {
+			return response.data.text[0];
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
@@ -82,7 +82,13 @@ async function translateGoogle(text, targetLang) {
 		}
 		throw new Error("Invalid response for translate request");
 	} catch (e) {
-		throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		if (e.response && e.response.status === 429) {
+			throw new Error(
+				`Could not translate to "${targetLang}": Rate-limited by Google Translate`
+			);
+		} else {
+			throw new Error(`Could not translate to "${targetLang}": ${e}`);
+		}
 	}
 }
 

